text/0225-configuration.md

```diff
@@ -0,0 +1,410 @@
+# OpenTelemetry Configuration
+
+A new configuration interface is proposed here in the form of a configuration model, which can be expressed as a file, and validated through a published schema.
+
+## Motivation
+
+OpenTelemetry specifies code that can operate in a variety of ways based on the end-user’s desired mode of operation. This requires a configuration interface be provided to the user so they are able to communicate this information. Currently, OpenTelemetry specifies this interface in the form of the API exposed by the SDKs and environment variables. This environment variable interface is limited in the structure of information it can communicate and the primitives it can support.
+
+### Environment Variable Interface Limitations
+
+The environment variable interface suffers from the following identified limitations:
+
+* **Flat**. Structured data is only allowed by using higher-level naming or data encoding schemes. Examples of configuration limited by lack of structured configuration include:
+  * Configuring multiple span processors, periodic metric readers, or log record processors.
+  * Configuring views.
+  * Configuring arguments for parent based sampler (sampler parent is remote and sampled vs. not sampled, sampler when parent is local and sampled vs. not sampled).
+* **Runtime dependent**. Different systems expose this interface differently (Linux, BSD, Windows). This usually means unique instructions are required to properly interact with the configuration interface on different systems.
+* **Limited values**. Many systems only allow string values to be used, but OpenTelemetry specifies many configuration values other than this type. For example, OTEL_RESOURCE_ATTRIBUTES specifies a list of key value pairs to be used as resource attributes, but there is no way to specify array values, or indicate that the value should be interpreted as non-string type.
+* **Limited validation**. Validation can only be performed by the receiver, there is no meta-configuration language to validate input.
+* **Difficult to extend**. It’s difficult to anticipate the requirements of configuring custom extension components (processors, exporters, samplers, etc), and likely not practical to represent them in a flat structure. As a result, the environment variable interface is limited to components explicitly mentioned in the specification.
+* **Lacks versioning**. The lack of versioning support for environment variables prevents evolution over time.
+
+## Explanation
+
+Using a configuration model or configuration file, users can configure all options currently available via environment variables.
+
+### Goals
```

Comment:
Two more goals that might be added or excluded explicitly:

* The use of multiple configuration files (like the collector allows it today)
* The possibility to include a configuration file into another one via an expression.

Reply:
> The use of multiple configuration files (like the collector allows it today)

Support for this is implied because you can programmatically load a config file and produce an SDK from it. Can do this multiple times if needed, with the caller responsible for passing the SDK instances to the appropriate places in the application.

> The possibility to include a configuration file into another one via an expression.

👎 I disagree. Layering config files is too complicated for the initial deliverable and would slow down a simpler solution that would work most of the time. Environment variable substitution can be used as a less powerful stand in.

Reply:
I've added "The possibility to include a configuration file into another one via an expression." as "Out of scope" under "goals" as I agree it adds a lot of complexity to the initial implementation.

> The use of multiple configuration files (like the collector allows it today)

@svrnm did you imagine this working as it does with the collector where the configuration is merged? if so then I would say that this is not implied, as repeatedly calling a function to load a config file doesn't imply that the resulting configured SDK is a merge of the previous calls.

Reply:
> did you imagine this working as it does with the collector where the configuration is merged

Thanks for clarifying - I wasn't aware the collector did this.

Reply:
> I've added "The possibility to include a configuration file into another one via an expression." as "Out of scope" under "goals" as I agree it adds a lot of complexity to the initial implementation.

I am OK with that, I just thought it would be good to clarify that.

> > The use of multiple configuration files (like the collector allows it today)
> 
> @svrnm did you imagine this working as it does with the collector where the configuration is merged? if so then I would say that this is not implied, as repeatedly calling a function to load a config file doesn't imply that the resulting configured SDK is a merge of the previous calls.

yes, this was what I was thinking about while reading through the OTEP. I just think about how end users interact with OpenTelemetry and if they see they can do it with a collector, they will definitely try to do the same with SDK configuration. Especially if a vendor-agnostic and vendor-specific config are supplied, this might become handy. Or, let's say you have a bunch of applications running on a k8s cluster and there is some generic configuration and some specific configuration you want to apply.